src/IECoreScene/SceneCache.cpp

```diff
@@ -728,8 +730,76 @@ class SceneCache::ReaderImplementation : public SceneCache::Implementation
 			return reader;
 		}
 
+		ConstPathMatcherDataPtr readSet( const Name &name ) const
+		{
+			IECore::PathMatcherDataPtr pathMatcherData = readLocalSet( name );
+
+			NameList children;
+			childNames( children );
+
+			for( const auto &c : children )
+			{
+				ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+				ConstPathMatcherDataPtr childSets = nc->child( c, SceneInterface::NullIfMissing )->readSet( name );
+				pathMatcherData->writable().addPaths(childSets->readable(), { c });
```

Comment:
This will add a level to PathMatcher data after we leave a recursion level which I think can be avoided.

Reply:
I'm not sure I understand. I think it's probably worth not using PathMatcherData at all during the recursion, as we're just making a wrapper that we throw away at each step. Instead we could return PathMatcher directly at each recursion and just wrap it in the outer layer. Perhaps that what you were getting at?

Reply:
> Instead we could return PathMatcher directly at each recursion and just wrap it in the outer layer. Perhaps that what you were getting at?

I don't think so. I was going to recurse to find all the PathMatchersData  and return a list of  [ ( prefix , PathMatcherData ), ... ] which I could perform PathMatcher.addPaths on the number of defined sets rather than depth I have to recurse to find a set. 

Or perahps I return nothing and pass the PathMatcher in the recursion and just add to it when I find a local set.

Reply:
> Or perahps I return nothing and pass the PathMatcher in the recursion and just add to it when I find a local set.

Sounds fine.

> I don't think so.

Obviously I haven't tested, but I would expect my suggestion to have comparable performance to the one you suggest above. In case it isn't obvious what I'm suggesting, I used the same approach recently in the [SceneGadget](https://github.com/GafferHQ/gaffer/blob/master/src/GafferSceneUI/SceneGadget.cpp#L354).